cmd/agent/api/internal/agent/agent.go

```diff
@@ -533,3 +540,62 @@ func max(a, b int) int {
 func GetConnection(r *http.Request) net.Conn {
 	return r.Context().Value(grpc.ConnContextKey).(net.Conn)
 }
+
+func dumpDogstatsdContexts(w http.ResponseWriter, _ *http.Request, demux demultiplexer.Component) {
+	if demux == nil {
+		setJSONError(w, log.Errorf("Unable to stream dogstatsd contexts, demultiplexer is not initialized"), 404)
+		return
+	}
+
+	path, err := dumpDogstatsdContextsImpl(demux)
+	if err != nil {
+		setJSONError(w, log.Errorf("Failed to create dogstatsd contexts dump: %v", err), 500)
+		return
+	}
+
+	resp, err := json.Marshal(path)
+	if err != nil {
+		setJSONError(w, log.Errorf("Failed to serialize response: %v", err), 500)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(resp)
+}
+
+func dumpDogstatsdContextsImpl(demux demultiplexer.Component) (string, error) {
+	path := path.Join(config.Datadog.GetString("run_path"), "dogstatsd_contexts.json.zstd")
+
+	f, err := os.Create(path)
+	if err != nil {
+		return "", err
+	}
+
+	c := zstd.NewWriter(f)
+
+	w := bufio.NewWriter(c)
+
+	err = demux.DumpDogstatsdContexts(w)
+	if err != nil {
+		return "", err
+	}
+
+	err = w.Flush()
+	if err != nil {
+		c.Close()
+		f.Close()
+		return "", err
+	}
+
+	err = c.Close()
+	if err != nil {
+		f.Close()
+		return "", err
+	}
+
+	err = f.Close()
+	if err != nil {
+		return "", err
+	}
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏really subjective but maybe the following would be clearer/less error prone
```suggestion
	for _, err := range []error{ w.Flush(), c.Close(), f.Close() } {
		if err != nil {
			return "", err
		}
	}
```